waku/node/v2/message_store.nim

```diff
@@ -0,0 +1,211 @@
+    if (let x = sqlite3_finalize(s); x != SQLITE_OK):
+      return err($sqlite3_errstr(x))
+
+  template checkExec(q: string) =
+    let s = prepare(q): discard
+    checkExec(s)
+
+  template checkWalPragmaResult(journalModePragma: ptr sqlite3_stmt) =
+    if (let x = sqlite3_step(journalModePragma); x != SQLITE_ROW):
+      discard sqlite3_finalize(journalModePragma)
+      return err($sqlite3_errstr(x))
+
+    if (let x = sqlite3_column_type(journalModePragma, 0); x != SQLITE3_TEXT):
+      discard sqlite3_finalize(journalModePragma)
+      return err($sqlite3_errstr(x))
+
+    if (let x = sqlite3_column_text(journalModePragma, 0);
+        x != "memory" and x != "wal"):
+      discard sqlite3_finalize(journalModePragma)
+      return err("Invalid pragma result: " & $x)
+
+  # TODO: check current version and implement schema versioning
+  checkExec "PRAGMA user_version = 1;"
+
+  let journalModePragma = prepare("PRAGMA journal_mode = WAL;"): discard
+  checkWalPragmaResult(journalModePragma)
+  checkExec(journalModePragma)
+
+  ## Table is the SQL query for creating the messages Table.
+  ## It contains:
+  ##  - 4-Byte ContentTopic stored as an Integer
+  ##  - Payload stored as a blob
+  checkExec """
+    CREATE TABLE IF NOT EXISTS messages (
+        id BLOB PRIMARY KEY,
+        contentTopic INTEGER NOT NULL, 
+        payload BLOB
+    ) WITHOUT ROWID;
+    """
+
+  ok(MessageStore(
+    env: env.release
+  ))
+
+template prepare(env: Sqlite, q: string, cleanup: untyped): ptr sqlite3_stmt =
+  var s: ptr sqlite3_stmt
+  checkErr sqlite3_prepare_v2(env, q, q.len.cint, addr s, nil):
+    cleanup
+  s
+
+proc bindParam(s: RawStmtPtr, n: int, val: auto): cint =
+  when val is openarray[byte]|seq[byte]:
+    if val.len > 0:
+      sqlite3_bind_blob(s, n.cint, unsafeAddr val[0], val.len.cint, nil)
+    else:
+      sqlite3_bind_blob(s, n.cint, nil, 0.cint, nil)
+  elif val is int32:
+    sqlite3_bind_int(s, n.cint, val)
+  elif val is uint32:
+    sqlite3_bind_int(s, int(n).cint, int(val).cint)
+  elif val is int64:
+    sqlite3_bind_int64(s, n.cint, val)
+  else:
+    {.fatal: "Please add support for the 'kek' type".}
+
+proc put*(db: MessageStore, message: WakuMessage): MessageStoreResult[void] =
+  let s = prepare(db.env, "INSERT INTO messages (id, contentTopic, payload) VALUES (?, ?, ?);"): discard
+  checkErr bindParam(s, 1, @(message.id().data))
+  checkErr bindParam(s, 2, message.contentTopic)
+  checkErr bindParam(s, 3, message.payload)
+
+  let res =
+    if (let v = sqlite3_step(s); v != SQLITE_DONE):
+      err($sqlite3_errstr(v))
+    else:
+      ok()
+
+  # release implict transaction
+  discard sqlite3_reset(s) # same return information as step
+  discard sqlite3_clear_bindings(s) # no errors possible
+
+  res
+
+proc close*(db: MessageStore) =
+  discard sqlite3_close(db.env)
+
+  db[] = MessageStore()[]
+
+proc get*(db: MessageStore, topics: seq[ContentTopic], paging: PagingInfo, onData: DataProc): MessageStoreResult[bool] =
```

Comment:
You will probably need to return a PaginInfo as part of this proc, and its elements must be set according to the pagination specs
its direction the same as the input paging
its pageSize is the length of the list of retrieved waku messages
its cursor is either the highest index (if direction is FORWARD) or the lowest index (if direction is BACKWARD) in the retrieved set

Reply:
I keep using index, it may be confusing, by index I mean the combination of timestamp and id. While comparing timestamp+id s, the timestamp is the most significant part

Reply:
@staheri14 can you have another look

Reply:
> You will probably need to return a PaginInfo as part of this proc, and its elements must be set according to the pagination specs
> its `direction` the same as the input `paging`
> its `pageSize` is the length of the list of retrieved waku messages
> its `cursor` is either the highest index (if direction is FORWARD) or the lowest index (if direction is BACKWARD) of the retrieved set

To clarify my comment, the `cursor` can be set as follows:
if direction == FORWARD then `cursor.digest `and `cursor.receivedTime` must be set to the `id` and `timestamp` of the last message in the retrieved page 
if direction == BACKWARD then `cursor.digest `and `cursor.receivedTime` must be set to the `id` and `timestamp` of the first message in the retrieved page